### PR TITLE
Jgfouca/branch for acme split 11 07 2017 pr

### PR DESCRIPTION
**Pull request by jgfouca**
Test suite: scripts_regression_tests
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes [CIME Github issue #]

User interface changes?: 

Update gh-pages html (Y/N)?:

Code review:

### CODE REVIEW

File: scripts/lib/CIME/SystemTests/pet.py [hunk 1]
```diff
@@ -33,6 +33,10 @@ def _case_one_setup(self):
         # Need to redo case_setup because we may have changed the number of threads
         case_setup(self._case, reset=True)
 
+        # Hack until CIME queue selection is more dynamic
```
Comment (jgfouca) [inline scripts/lib/CIME/SystemTests/pet.py]: This shouldn't have been re-introduced.

File: scripts/lib/CIME/SystemTests/system_tests_compare_two.py [hunk 2]
```diff
@@ -304,7 +304,7 @@ def _get_case2_exeroot(self):
             # to this case. (If case2's CIME_OUTPUT_ROOT were in some
             # more generic location, then this would result in its bld
             # directory being inadvertently shared with other tests.)
-            case2_exeroot = os.path.join(self._get_output_root2(), "bld")
+            case2_exeroot = os.path.join(self._get_output_root2(), self._case1.get_value("CASE"), "bld")
```
Comment (jgfouca) [inline scripts/lib/CIME/SystemTests/system_tests_compare_two.py]: Shouldn't have been reintroduced.

File: scripts/lib/CIME/XML/env_mach_specific.py [hunk 3]
```diff
@@ -222,12 +222,12 @@ def _match_attribs(self, attribs, case):
 
     def _match(self, my_value, xml_value):
         if xml_value.startswith("!"):
-            result = my_value != xml_value[1:]
+            result = re.match(xml_value[1:],str(my_value)) is None
```
Comment (jgfouca) [inline scripts/lib/CIME/XML/env_mach_specific.py]: Important change.

File: scripts/lib/CIME/case_run.py [hunk 4]
```diff
@@ -106,7 +106,11 @@ def _run_model_impl(case, lid, skip_pnl=False, da_cycle=0):
 
     while loop:
         loop = False
-        stat = run_cmd(cmd, from_dir=rundir)[0]
+
+        save_prerun_provenance(case)
```
Comment (jgfouca) [inline scripts/lib/CIME/case_run.py]: Important change.

File: scripts/lib/CIME/case_st_archive.py [hunk 5]
```diff
@@ -143,11 +143,13 @@ def _datetime_str_mpas(date):
     '0011-12-09_00435'
     """
 
-    format_string = "{year:04d}-{month:02d}-{day:02d}_{seconds:05d}"
+    format_string = "{year:04d}-{month:02d}-{day:02d}_{hours:02d}:{minutes:02d}:{seconds:02d}"
```
Comment (jgfouca) [inline scripts/lib/CIME/case_st_archive.py]: @mfdeakin-sandia ?

File: scripts/lib/CIME/provenance.py [hunk 6]
```diff
@@ -191,8 +191,28 @@ def _save_prerun_timing_acme(case, lid):
         for item in glob.glob(os.path.join(blddir, blddir_glob_to_copy)):
             copy_umask(item, os.path.join(full_timing_dir, os.path.basename(item) + "." + lid))
 
+    # Save state of repo
+    if os.path.exists(os.path.join(cimeroot, ".git")):
```
Comment (jgfouca) [inline scripts/lib/CIME/provenance.py]: Only impacts ACME.

File: scripts/lib/CIME/wait_for_tests.py [hunk 7]
```diff
@@ -52,7 +52,15 @@ def get_test_output(test_path):
 ###############################################################################
 def create_cdash_test_xml(results, cdash_build_name, cdash_build_group, utc_time, current_time, hostname):
 ###############################################################################
-    git_commit = CIME.utils.get_current_commit(repo=CIME.utils.get_cime_root())
+    # We assume all cases were created from the same code repo
+    first_result_case = os.path.dirname(results.iteritems().next()[1][0])
```
Comment (jgfouca) [inline scripts/lib/CIME/wait_for_tests.py]: Only affects acme.

File: src/drivers/mct/cime_config/config_component.xml [hunk 8]
```diff
@@ -510,6 +510,24 @@
     If TRUE, short term archiving will be turned on.</desc>
   </entry>
 
+  <entry id="DOUT_L_MS">
+    <type>logical</type>
+    <valid_values>TRUE,FALSE</valid_values>
+    <default_value>FALSE</default_value>
+    <group>run_data_archive</group>
```
Comment (jgfouca) [inline src/drivers/mct/cime_config/config_component.xml]: This should not have been reintroduced.

File: src/drivers/mct/cime_config/config_component.xml [hunk 9]
```diff
@@ -2282,15 +2300,15 @@
...
-      <value compclass="ATM">1</value>
-      <value compclass="CPL">1</value>
-      <value compclass="OCN">1</value>
-      <value compclass="WAV">1</value>
-      <value compclass="GLC">1</value>
-      <value compclass="ICE">1</value>
-      <value compclass="ROF">1</value>
-      <value compclass="LND">1</value>
-      <value compclass="ESP">1</value>
+      <value component="ATM">0</value>
```
Comment (jgfouca) [inline src/drivers/mct/cime_config/config_component.xml]: This seems important.

File: src/externals/pio1/tests/testpio/perl5lib/XML/man3/XML_Lite.3 [hunk 10]
```diff
@@ -0,0 +1,213 @@
+.\" Automatically generated by Pod::Man version 1.02
+.\" Tue Mar 18 09:37:35 2003
+.\"
+.\" Standard preamble:
+.\" ======================================================================
```
Comment (jgfouca) [inline src/externals/pio1/tests/testpio/perl5lib/XML/man3/XML_Lite.3]: More zombies?

File: src/drivers/mct/main/cime_comp_mod.F90 [hunk 11]
```diff
@@ -422,6 +422,7 @@ module cime_comp_mod
    logical  :: reprosum_recompute     ! setup reprosum, recompute if tolerance exceeded
 
    logical  :: output_perf = .false.  ! require timing data output for this pe
+   logical  :: in_first_day = .true.  ! currently simulating first day
```
Comment (jgfouca) [inline src/drivers/mct/main/cime_comp_mod.F90]: @rljacob , are these driver changes legit?

File: scripts/lib/CIME/case_st_archive.py (hunk 5 above)
Comment (mfdeakin-sandia) [reply scripts/lib/CIME/case_st_archive.py]: Yes, this change is correct for MPAS rpointer files

File: scripts/lib/CIME/case_run.py [hunk 12]
```diff
@@ -275,21 +279,22 @@ def case_run(case, skip_pnl=False):
...
-        case.read_xml()
-
     for cycle in range(data_assimilation_cycles):
         # After the first DA cycle, runs are restart runs
         if cycle > 0:
             case.set_value("CONTINUE_RUN", "TRUE")
             lid = new_lid()
 
+        if prerun_script:
+            case.flush()
```
Comment (jgfouca) [inline scripts/lib/CIME/case_run.py]: Important change.

**Comment by rljacob** [comment]
How many weeks is this? 106 commits seems like a lot from ACME.

**Comment by jgfouca** [comment]
About two.

**Comment by jgfouca** [comment]
All, this is ready for review. All tests are passing and the changes have been annotated.

File: src/drivers/mct/cime_config/config_component.xml [hunk 13]
```diff
@@ -510,6 +510,24 @@
     If TRUE, short term archiving will be turned on.</desc>
   </entry>
 
+  <entry id="DOUT_L_MS">
```
Comment (jedwards4b) [inline src/drivers/mct/cime_config/config_component.xml]: Another regression - this code should be removed.

File: src/drivers/mct/cime_config/config_component.xml (hunk 9 above)
Comment (jedwards4b) [inline src/drivers/mct/cime_config/config_component.xml]: Why this change?   Should be made in an acme specific way such as in config/machines/config_pio.xml

File: src/drivers/mct/cime_config/config_component_acme.xml [hunk 14]
```diff
@@ -255,16 +255,16 @@
...
-      <value compset="_DATM.*_SLND.*MPASO" grid="oi%oQU240wLI">12</value>
-      <value compset="_DATM.*_SLND.*MPASO" grid="oi%oQU120">24</value>
-      <value compset="_DATM.*_SLND.*MPASO" grid="oi%oEC60to30">48</value>
-      <value compset="_DATM.*_SLND.*MPASO" grid="oi%oEC60to30v3">48</value>
-      <value compset="_DATM.*_SLND.*MPASO" grid="oi%oEC60to30wLI">48</value>
-      <value compset="_DATM.*_SLND.*MPASO" grid="oi%oRRS30to10v3">96</value>
-      <value compset="_DATM.*_SLND.*MPASO" grid="oi%oRRS30to10wLI">96</value>
-      <value compset="_DATM.*_SLND.*MPASO" grid="oi%oRRS18to6v3">96</value>
-      <value compset="_DATM.*_SLND.*MPASO" grid="oi%oRRS15to5">96</value>
+      <value compset="_DATM.*_SLND.*MPAS" grid="oi%oQU240">12</value>
```
Comment (jedwards4b) [inline src/drivers/mct/cime_config/config_component_acme.xml]: Why was the "O" removed - seems like a source of conflict with MPASLI or MPASA

File: src/externals/pio1/tests/testpio/perl5lib/XML/man3/XML_Lite.3 [hunk 15]
```diff
@@ -0,0 +1,213 @@
+.\" Automatically generated by Pod::Man version 1.02
```
Comment (jedwards4b) [inline src/externals/pio1/tests/testpio/perl5lib/XML/man3/XML_Lite.3]: regresssion - this file should not be here.

File: src/externals/pio1/tests/testpio/perl5lib/XML/man3/XML_Lite_Element.3 [hunk 16]
```diff
@@ -0,0 +1,206 @@
+.\" Automatically generated by Pod::Man version 1.02
```
Comment (jedwards4b) [inline src/externals/pio1/tests/testpio/perl5lib/XML/man3/XML_Lite_Element.3]: remove again

File: src/share/timing/perf_mod.F90 [hunk 17]
```diff
@@ -839,6 +841,95 @@ end subroutine t_stopf
...
+!pw      ierr = GPTLstartstop_vals(event, wtime)
+!pw   endif
+
+   endif
+
+   return
+   end subroutine t_startstop_valsf
+!
+!========================================================================
+!
```
Comment (jedwards4b) [inline src/share/timing/perf_mod.F90]: Significant changes to the timing library and output were made.  I have no specific objections as to the content of the changes, but find the process more than a little frustrating - there should be issues, discussion and pr's related to these changes.

File: src/drivers/mct/cime_config/config_component.xml (hunk 9 above)
Comment (jgfouca) [reply src/drivers/mct/cime_config/config_component.xml]: @jayeshkrishna ?

File: scripts/lib/CIME/XML/env_mach_specific.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgfouca) [reply scripts/lib/CIME/XML/env_mach_specific.py]: I think @amametjanov made this change so he could match grids/compsets without having to use the full name.

File: src/drivers/mct/cime_config/config_component_acme.xml (hunk 14 above)
Comment (jgfouca) [reply src/drivers/mct/cime_config/config_component_acme.xml]: Beats me. @amametjanov ?

Comment (jonbob) [reply src/drivers/mct/cime_config/config_component_acme.xml]: Those are mpas-o/mpas-cice grids. We would want those coupling intervals for compsets that are datm and either of the mpas ocn/ice components

File: scripts/lib/CIME/case.py [hunk 18]
```diff
@@ -139,6 +139,8 @@ def initialize_derived_attributes(self):
             "threaded" : self.get_build_threaded(),
             }
 
+        os.environ["OMP_NUM_THREADS"] = str(self.thread_count)
```
[the remark this answers is not in the document]
Comment (jgfouca) [reply scripts/lib/CIME/case.py]: @amametjanov ?

**Comment by jgfouca** [comment]
@jedwards4b , I think a significant number of your comments were on code that I already fixed. Can you look at case_run again?

**Comment by jedwards4b** [comment]
@jgfouca Yes I saw the update, looks better, thanks.

File: scripts/Tools/case_diff [hunk 19]
```diff
@@ -132,7 +132,10 @@ def _main_func(description):
     for xml_normalize_field in xml_normalize_fields:
         val1 = run_cmd_no_fail("./xmlquery --value {}".format(xml_normalize_field), from_dir=case1)
         val2 = run_cmd_no_fail("./xmlquery --value {}".format(xml_normalize_field), from_dir=case2)
-        repls[val2] = val1
+        if "/" in val1:
```
[the remark this answers is not in the document]
Comment (jgfouca) [reply scripts/Tools/case_diff]: Done.

File: config/xml_schemas/env_mach_specific.xsd [hunk 20]
```diff
@@ -81,20 +81,10 @@
...
-      <xs:sequence>
-        <xs:element maxOccurs="unbounded" ref="env"/>
-      </xs:sequence>
-      <xs:attribute ref="DEBUG"/>
-      <xs:attribute ref="mpilib"/>
-      <xs:attribute ref="compiler"/>
-      <xs:attribute ref="unit_testing"/>
-      <xs:attribute ref="SMP_PRESENT"/>
-    </xs:complexType>
+  <xs:element name="environment_variables" type="xs:anyType">
```
[the remark this answers is not in the document]
Comment (jgfouca) [reply config/xml_schemas/env_mach_specific.xsd]: @jedwards4b , i'm not very good with XSD. Can you recommend something better?

File: config/xml_schemas/config_machines.xsd [hunk 21]
```diff
@@ -192,10 +192,7 @@
       <xs:sequence>
         <xs:element maxOccurs="unbounded" ref="env"/>
       </xs:sequence>
-      <xs:attribute ref="DEBUG"/>
-      <xs:attribute ref="mpilib"/>
-      <xs:attribute ref="compiler"/>
-      <xs:attribute ref="unit_testing"/>
+      <xs:anyAttribute/>
```
Comment (jedwards4b) [inline config/xml_schemas/config_machines.xsd]: I think that this will meet the requirement.

File: config/xml_schemas/env_mach_specific.xsd [hunk 22]
```diff
@@ -81,19 +81,15 @@
...
     <xs:complexType>
       <xs:sequence>
         <xs:element maxOccurs="unbounded" ref="env"/>
       </xs:sequence>
-      <xs:attribute ref="DEBUG"/>
-      <xs:attribute ref="mpilib"/>
-      <xs:attribute ref="compiler"/>
-      <xs:attribute ref="unit_testing"/>
-      <xs:attribute ref="SMP_PRESENT"/>
+      <xs:anyAttribute/>
```
Comment (jedwards4b) [inline config/xml_schemas/env_mach_specific.xsd]: I think that this is what is needed.

File: scripts/lib/CIME/case.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (amametjanov) [reply scripts/lib/CIME/case.py]: @jedwards4b this is needed to be able to query number of threads at setup- and build-time: e.g. at `./preview_run`

[the remark this answers is not in the document]
Comment (jedwards4b) [reply scripts/lib/CIME/case.py]: Why not use self.thread_count (case.thread_count) directly then?

File: src/drivers/mct/cime_config/config_component_acme.xml (hunk 14 above)
Comment (jedwards4b) [reply src/drivers/mct/cime_config/config_component_acme.xml]: You should consider using MPAS[IO] then.

File: config/xml_schemas/env_mach_specific.xsd (hunk 20 above)
[the remark this answers is not in the document]
Comment (amametjanov) [reply config/xml_schemas/env_mach_specific.xsd]: This was needed to allow any case-var as an attribute on that block. IIRC, anyAttribute still requires those attributes to be declared: i.e.
```
element environment_variables: Schemas validity error : Element 'environment_variables', attribute 'MAX_MPITASKS_PER_NODE': No matching global attribute declaration available, but demanded by the strict wildcard.
```

File: src/drivers/mct/cime_config/config_component_acme.xml (hunk 14 above)
Comment (jonbob) [reply src/drivers/mct/cime_config/config_component_acme.xml]: That's fine with me.

File: scripts/lib/CIME/case.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (amametjanov) [reply scripts/lib/CIME/case.py]: Outside of python, at the shell prompt, running `./preview_run` leaves $ENV{OMP_NUM_THREADS} unresolved.

[the remark this answers is not in the document]
Comment (jedwards4b) [reply scripts/lib/CIME/case.py]: This would seem to be a deficiency in preview_run then.  OMP_NUM_THREADS should not need to be set here.

File: scripts/lib/CIME/XML/env_mach_specific.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jedwards4b) [reply scripts/lib/CIME/XML/env_mach_specific.py]: I'm not sure I understand the implications of this, if you can match without a full name it seems like you can match the wrong thing.

**Comment by jedwards4b** [comment]
@amametjanov I think I've addressed the xsd issue, please confirm.

File: src/share/timing/perf_mod.F90 (hunk 17 above)
Comment (worleyph) [reply src/share/timing/perf_mod.F90]: @jgfouca , yes, these are my changes. And there are more than just fortran changes (also C changes to gptl.c etc.) There were issues and PRs on the ACME side. Wanted to capture time spent prior to the call to initialize the timing library, since MPI_Init can cost as much as 20 minutes on Cori-KNL in some circumstances. Other tweaks were to: (a) more consistently add CPL: to cpl-related timers. Seemed an obvious clean-up, and I did not apply it everywhere, since it was delberately left off of a few timers and I do not know the history to these decisions, (b) force a checkpoint after the first simulated day, so that long initialization times can be captured even if a job runs out of wallclock time, (c) slightly widen the scope of the prefix application in component_init_cc, as this proved useful in performance debugging, and (d) do not add negative times to running totals (when the underlying timers is not monotonic).

File: config/xml_schemas/env_mach_specific.xsd [hunk 23]
```diff
@@ -89,7 +89,7 @@
       <xs:sequence>
         <xs:element maxOccurs="unbounded" ref="env"/>
       </xs:sequence>
-      <xs:anyAttribute/>
+      <xs:anyAttribute processContents="skip"/>
```
Comment (amametjanov) [inline config/xml_schemas/env_mach_specific.xsd]: This seems to work.

File: scripts/lib/CIME/case.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (amametjanov) [reply scripts/lib/CIME/case.py]: Is there a specific reason why it should not be set here? `OMP_NUM_THREADS` is known right after `case.setup`. Having it captured prior to building in things like build_environment also helps in debugging.

File: src/drivers/mct/cime_config/config_component.xml (hunk 9 above)
Comment (jgfouca) [reply src/drivers/mct/cime_config/config_component.xml]: I will put this in config_pio.xml, do it for all machines.

File: src/drivers/mct/cime_config/config_component_acme.xml (hunk 14 above)
Comment (jgfouca) [reply src/drivers/mct/cime_config/config_component_acme.xml]: @jonbob, please make the change suggested by JimE on the ACME side if you want it.

File: scripts/lib/CIME/case.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (jgfouca) [reply scripts/lib/CIME/case.py]: Will leave as-is for now. Not doing any harm.

File: scripts/lib/CIME/XML/env_mach_specific.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgfouca) [reply scripts/lib/CIME/XML/env_mach_specific.py]: @amametjanov , we need to back-out this change and figure out why pio.py was not already using regex matching.

[the remark this answers is not in the document]
Comment (amametjanov) [reply scripts/lib/CIME/XML/env_mach_specific.py]: This was not for PIO, but for matching to case-var values. For example, an attribute could be `GRID` and its value can be very long:
```
$ ./xmlquery --value GRID
a%ne120np4_l%ne120np4_oi%oRRS18to6v3_ICG_r%r0125_m%oRRS18to6v3_g%null_w%null
```
Having this, enables pattern-matching like
```
<environment_variables GRID=a%ne120.* >
...
```

[the remark this answers is not in the document]
Comment (jedwards4b) [reply scripts/lib/CIME/XML/env_mach_specific.py]: Okay - I understand now, @jgfouca let's leave it.

File: config/acme/machines/config_pio.xml [hunk 24]
```diff
@@ -28,13 +28,11 @@
     </values>
   </entry>
 
-  <!--- uncomment and fill in relevant sections
   <entry id="PIO_ROOT">
     <values>
-      <value></value>
```
Comment (rljacob) [inline config/acme/machines/config_pio.xml]: This is enough to change each model's ROOT?

Comment (jgfouca) [reply config/acme/machines/config_pio.xml]: yes, I confirmed it with a quick test

```
./xmlquery PIO_ROOT
	PIO_ROOT: ['CPL:0', 'ATM:0', 'LND:0', 'ICE:0', 'OCN:0', 'ROF:0', 'GLC:0', 'WAV:0', 'ESP:0']
```
